tools/dataflow-throttling/README.md

```diff
@@ -0,0 +1,116 @@
+# Client-side Throttling in Dataflow
+
+### Introduction
+When Dataflow pipeline is trying to get the response for each incoming event from third-party API’s there is a chance of the request to fail in getting response for some reasons like network issues, limits on API calls, etc. At times, client-side throttling can act as a way to limit the number of requests coming from dataflow by either postponing or dropping excessive requests. It can also serve as a fail-safe mechanism that prevents the loss of payload of requests throttled by the server.
+
+For example, when a streaming pipeline calls the Cloud DLP API for each event to inspect sensitive data or de-identify the additional columns, DLP API can accept only 600 requests per minute though the incoming flow can be more than the limit (this quota can be increased from the GCP console). In this scenario, Instead of sending events to DLP API and getting failures, client-side throttling buffers the requests locally and fails them only if the DLP service starts to throttle the client even if the number of requests sent by Dataflow is below quota. Dataflow throttling can also reroute the events to a Dead Letter Queue or to any other sink from where we can reprocess the events.
+
+This artifact is designed to implement distributed adaptive client-side throttling. [Adaptive throttling](https://landing.google.com/sre/sre-book/chapters/handling-overload/#eq2101) activates when a service called by Dataflow starts rejecting requests due to throughput higher than expected. This transform tracks the request rejection probability to decide whether it should fail locally without going through network to save the cost of rejecting a request.
```

Comment:
This transform -> Transform class from this library

Reply:
This transform tracks the request rejection probability to decide whether it should fail locally without going through network to save the cost of rejecting a request. -> The main transform in this library tracks the request rejection probability to decide whether it should fail locally without going through network to save the cost of rejecting a request on the client and server sides.